README.md
dalek-reporter-json
===================

> DalekJS reporter plugin for JSON output

[![Build Status](https://travis-ci.org/dalekjs/dalek-reporter-json.png)](https://travis-ci.org/dalekjs/dalek-reporter-json)
[![Build Status](https://drone.io/github.com/dalekjs/dalek-reporter-json/status.png)](https://drone.io/github.com/dalekjs/dalek-reporter-json/latest)
[![Dependency Status](https://david-dm.org/dalekjs/dalek-reporter-json.png)](https://david-dm.org/dalekjs/dalek-reporter-json)
[![devDependency Status](https://david-dm.org/dalekjs/dalek-reporter-json/dev-status.png)](https://david-dm.org/dalekjs/dalek-reporter-json#info=devDependencies)
[![NPM version](https://badge.fury.io/js/dalek-reporter-json.png)](http://badge.fury.io/js/dalek-reporter-json)
[![Coverage](http://dalekjs.com/package/dalek-reporter-json/master/coverage/coverage.png)](http://dalekjs.com/package/dalek-reporter-json/master/coverage/index.html)
[![unstable](https://rawgithub.com/hughsk/stability-badges/master/dist/unstable.svg)](http://github.com/hughsk/stability-badges)
[![Stories in Ready](https://badge.waffle.io/dalekjs/dalek.png?label=ready)](https://waffle.io/dalekjs/dalek-reporter-json)
[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/dalekjs/dalek-reporter-json/trend.png)](https://bitdeli.com/free "Bitdeli Badge")
[![Built with Grunt](https://cdn.gruntjs.com/builtwith.png)](http://gruntjs.com/)

[![NPM](https://nodei.co/npm/dalek-reporter-json.png)](https://nodei.co/npm/dalek-reporter-json/)
[![NPM](https://nodei.co/npm-dl/dalek-reporter-json.png)](https://nodei.co/npm/dalek-reporter-json/)

## Ressources

[API Docs](http://dalekjs.com/package/dalek-reporter-json/master/api/index.html) -
[Trello](https://trello.com/b/wpindMWl/dalek-reporter-json) -
[Code coverage](http://dalekjs.com/package/dalek-reporter-json/master/coverage/index.html) -
[Code complexity](http://dalekjs.com/package/dalek-reporter-json/master/complexity/index.html) -
[Contributing](https://github.com/dalekjs/dalek-reporter-json/blob/master/CONTRIBUTING.md) -
[User Docs](http://dalekjs.com/docs/json.html) -
[Homepage](http://dalekjs.com) -
[Twitter](http://twitter.com/dalekjs)

## Docs

The JSON reporter can produce a file with the results of your testrun.

The reporter can be installed with the following command:
```
$ npm install dalek-reporter-json --save-dev
```

The file will follow the following format. This is a first draft and will
definitly change in future versions.

```javascript
{
  "tests": [
        {
          "id": "test806",
          "name": "Can get !url (OK, TDD style, message, chained)",
          "browser": "Chrome",
          "status": true,
          "passedAssertions": 1,
          "failedAssertions": 0,
          "actions": [
              {
                  "value": "http://localhost:5000/index.html",
                  "type": "open",
                  "uuid": "6ea84fc0-58bf-4e1f-bb9c-f035c6e6fae2",
                  "kind": "action",
                  "isAction": true
              },
              {
                  "success": true,
                  "expected": "http://localhost:5000/guinea.html",
                  "value": "http://localhost:5000/index.html",
                  "message": "Url is not whatever",
                  "type": "url",
                  "kind": "assertion",
                  "isAssertion": true
              }
          ]
      }
  ],
  "elapsedTime": {
      "minutes": 1,
      "seconds": 43.328535046
  },
  "status": true,
  "assertions": 1,
  "assertionsFailed": 0,
  "assertionsPassed": 1
}
```

By default the file will be written to `report/dalek.json`,
you can change this by adding a config option to the your Dalekfile

```javascript
"json-reporter": {
  "dest": "your/folder/your_file.json"
}
```

## Help Is Just A Click Away

### #dalekjs on FreeNode.net IRC

Join the `#daleksjs` channel on [FreeNode.net](http://freenode.net) to ask questions and get help.

### [Google Group Mailing List](https://groups.google.com/forum/#!forum/dalekjs)

Get announcements for new releases, share your projects and ideas that are
using DalekJS, and join in open-ended discussion that does not fit in
to the Github issues list or StackOverflow Q&A.

**For help with syntax, specific questions on how to implement a feature
using DalekJS, and other Q&A items, use StackOverflow.**

### [StackOverflow](http://stackoverflow.com/questions/tagged/dalekjs)

Ask questions about using DalekJS in specific scenarios, with
specific features. For example, help with syntax, understanding how a feature works and
how to override that feature, browser specific problems and so on.

Questions on StackOverflow often turn in to blog posts or issues.

### [Github Issues](//github.com/dalekjs/dalek-reporter-json/issues)

Report issues with DalekJS, submit pull requests to fix problems, or to
create summarized and documented feature requests (preferably with pull
requests that implement the feature).

**Please don't ask questions or seek help in the issues list.** There are
other, better channels for seeking assistance, like StackOverflow and the
Google Groups mailing list.

![DalekJS](https://raw.github.com/dalekjs/dalekjs.com/master/img/logo.png)

## Legal FooBar (MIT License)

Copyright (c) 2013 Sebastian Golasch

Distributed under [MIT license](https://github.com/dalekjs/dalek-reporter-json/blob/master/LICENSE-MIT)

CHANGELOG.md
<a name="0.0.1"></a>
### 0.0.1 (2014-11-28)


#### Bug Fixes

* **core:** linting errors ([0ce34bed](http://github.com/dalekjs/dalek-reporter-json/commit/0ce34bed7532c3a0d469a702920f73e10025db38))




test/index_TEST.js
'use strict';

var expect = require('chai').expect;

describe('dalek-reporter-json', function() {

  it('should be ok', function () {
    var JSONReporter = require('../index')({
      events: {emit: function () {}, on: function () {}, off: function () {}},
      config: {get: function () {}}
    });
    expect(JSONReporter).to.be.ok;
  });


  it('can store browser string', function () {
    var JSONReporter = require('../index')({
      events: {emit: function () {}, on: function () {}, off: function () {}},
      config: {get: function () {}}
    });

    JSONReporter.runBrowser('Chrome');
    expect(JSONReporter.browser).to.equal('Chrome');

    JSONReporter.runBrowser('PhantomJS');
    expect(JSONReporter.browser).to.equal('PhantomJS');
  });

  it('can store action', function () {
    var JSONReporter = require('../index')({
      events: {emit: function () {}, on: function () {}, off: function () {}},
      config: {get: function () {}}
    });

    JSONReporter.actionQueue = [];
    JSONReporter.action({name: 'an action', type: 'val'});
    expect(JSONReporter.actionQueue[0]).to.deep.equal({name: 'an action', type: 'val', kind: 'action'});
  });

  it('can store assertions', function () {
    var JSONReporter = require('../index')({
      events: {emit: function () {}, on: function () {}, off: function () {}},
      config: {get: function () {}}
    });

    JSONReporter.actionQueue = [];
    JSONReporter.assertion({name: 'an assertions', type: 'val'});
    expect(JSONReporter.actionQueue[0]).to.deep.equal({name: 'an assertions', type: 'val', kind: 'assertion'});
  });

  it('can store testStarted information', function () {
    var JSONReporter = require('../index')({
      events: {emit: function () {}, on: function () {}, off: function () {}},
      config: {get: function () {}}
    });

    JSONReporter.actionQueue = ['foobar'];
    JSONReporter.testStarted({name: 'xxxx', type: 'yyyy'});
    expect(JSONReporter.actionQueue).to.be.empty;
    expect(JSONReporter.currentTest).to.deep.equal({name: 'xxxx', type: 'yyyy'});
  });

  it('can store testFinished information', function () {
    var JSONReporter = require('../index')({
      events: {emit: function () {}, on: function () {}, off: function () {}},
      config: {get: function () {}}
    });

    JSONReporter.runBrowser('Chrome');
    JSONReporter.actionQueue = [];
    JSONReporter.testFinished({id: '1', name: 'val', status: true, passedAssertions: 1, failedAssertions: 0});
    expect(JSONReporter.data.tests[0]).to.deep.equal({id: '1', name: 'val', status: true, passedAssertions: 1, failedAssertions: 0, browser: 'Chrome', actions: []});
  });

});


Gruntfile.js
module.exports = function (grunt) {
  'use strict';

  // check task runtime
  require('time-grunt')(grunt);

  // load generic configs
  var configs = require('dalek-build-tools');

  grunt.initConfig({

    // load module meta data
    pkg: grunt.file.readJSON('package.json'),

    // define a src set of files for other tasks
    src: {
      lint: ['Gruntfile.js', 'index.js', 'test/*.js'],
      complexity: ['index.js'],
      test: ['test/*.js'],
      src: ['index.js']
    },

    // clean automatically generated helper files & docs
    clean: configs.clean,

    // speed up build by defining concurrent tasks
    concurrent: configs.concurrent,

    // linting
    jshint: configs.jshint,

    // testing
    mochaTest: configs.mocha,

    // code metrics
    complexity: configs.complexity,
    plato: configs.plato(grunt.file.readJSON('.jshintrc')),

    // api docs
    yuidoc: configs.yuidocs(),

    // user docs
    documantix: {
      options: {
        header: 'dalekjs/dalekjs.com/master/assets/header.html',
        footer: 'dalekjs/dalekjs.com/master/assets/footer.html',
        target: 'report/docs',
        vars: {
          title: 'DalekJS - Documentation - Reporter - JSON',
          desc: 'DalekJS - Documentation - Reporter - JSON',
          docs: true
        }
      },
      src: ['index.js']
    },

    // add current timestamp to the html document
    includereplace: {
      dist: {
        options: {
          globals: {
            timestamp: '<%= grunt.template.today("dddd, mmmm dS, yyyy, h:MM:ss TT") %>'
          },
        },
        src: 'report/docs/*.html',
        dest: './'
      }
    },

    // up version, tag & commit
    bump: configs.bump({
      pushTo: 'git@github.com:dalekjs/dalek-reporter-json.git',
      files: ['package.json', 'CONTRIBUTORS.md', 'CHANGELOG.md']
    }),

    // generate contributors file
    contributors: configs.contributors,

    // compress artifacts
    compress: configs.compress,

    // prepare files for grunt-plato to
    // avoid error messages (weird issue...)
    preparePlato: {
      options: {
        folders: [
          'coverage',
          'report',
          'report/coverage',
          'report/complexity',
          'report/complexity/files',
          'report/complexity/files/index_js'
        ],
        files: [
          'report.history.json',
          'files/index_js/report.history.json'
        ]
      }
    },

    // prepare files & folders for coverage
    prepareCoverage: {
      options: {
        folders: ['coverage', 'report', 'report/coverage'],
        pattern: '[require("fs").realpathSync(__dirname + "/../index.js")]'
      }
    },

    // archive docs
    archive: {
      options: {
        files: ['json.html']
      }
    },

    // release canary version
    'release-canary': {
      options: {
        files: []
      }
    }

  });
  // load 3rd party tasks
  require('load-grunt-tasks')(grunt);
  grunt.loadTasks('./node_modules/dalek-build-tools/tasks');

  // define runner tasks
  grunt.registerTask('lint', 'jshint');
  
  // split test & docs for speed
  grunt.registerTask('test', ['clean:coverage', 'prepareCoverage', 'concurrent:test', 'generateCoverageBadge']);
  grunt.registerTask('docs', ['clean:reportZip', 'clean:report', 'preparePlato', 'concurrent:docs', 'documantix', 'includereplace', 'compress']);
  
  // release tasks
  grunt.registerTask('releasePatch', ['test', 'bump-only:patch', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMinor', ['test', 'bump-only:minor', 'contributors', 'changelog', 'bump-commit']);
  grunt.registerTask('releaseMajor', ['test', 'bump-only:major', 'contributors', 'changelog', 'bump-commit']);
  
  // clean, test, generate docs (the CI task)
  grunt.registerTask('all', ['clean', 'test', 'docs']);
};


index.js
/*!
 *
 * Copyright (c) 2013 Sebastian Golasch
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 */

'use strict';

// ext. libs
var fs = require('fs');
var path = require('path');

// int. global
var reporter = null;

/**
 * The JSON reporter can produce a file with the results of your testrun.
 *
 * The reporter can be installed with the following command:
 * ```
 * $ npm install dalek-reporter-json --save-dev
 * ```
 *
 * The file will follow the following format. This is a first draft and will
 * definitly change in future versions.
 *
 * ```javascript
 * {
 *   "tests": [
 *         {
 *           "id": "test806",
 *           "name": "Can get !url (OK, TDD style, message, chained)",
 *           "browser": "Chrome",
 *           "status": true,
 *           "passedAssertions": 1,
 *           "failedAssertions": 0,
 *           "actions": [
 *               {
 *                   "value": "http://localhost:5000/index.html",
 *                   "type": "open",
 *                   "uuid": "6ea84fc0-58bf-4e1f-bb9c-f035c6e6fae2",
 *                   "kind": "action",
 *                   "isAction": true
 *               },
 *               {
 *                   "success": true,
 *                   "expected": "http://localhost:5000/guinea.html",
 *                   "value": "http://localhost:5000/index.html",
 *                   "message": "Url is not whatever",
 *                   "type": "url",
 *                   "kind": "assertion",
 *                   "isAssertion": true
 *               }
 *           ]
 *       }
 *   ],
 *   "elapsedTime": {
 *       "minutes": 1,
 *       "seconds": 43.328535046
 *   },
 *   "status": true,
 *   "assertions": 1,
 *   "assertionsFailed": 0,
 *   "assertionsPassed": 1
 * }
 * ```
 *
 * By default the file will be written to `report/dalek.json`,
 * you can change this by adding a config option to the your Dalekfile
 *
 * ```javascript
 * "json-reporter": {
 *   "dest": "your/folder/your_file.json"
 * }
 * ```
 *
 * @class Reporter
 * @constructor
 * @part JSON
 * @api
 */

function Reporter (opts) {
  this.events = opts.events;
  this.config = opts.config;
  this.data = {};
  this.actionQueue = [];
  this.data.tests = [];
  this.browser = null;

  var defaultReportFolder = 'report';
  this.dest = this.config.get('json-reporter') && this.config.get('json-reporter').dest ? this.config.get('json-reporter').dest : defaultReportFolder;

  this.startListening();
}

/**
 * @module Reporter
 */

module.exports = function (opts) {
  if (reporter === null) {
    reporter = new Reporter(opts);
  }

  return reporter;
};

Reporter.prototype = {

  /**
   * Connects to all the event listeners
   *
   * @method startListening
   * @chainable
   */

  startListening: function () {
    this.events.on('report:run:browser', this.runBrowser.bind(this));
    this.events.on('report:assertion', this.assertion.bind(this));
    this.events.on('report:action', this.action.bind(this));
    this.events.on('report:test:started', this.testStarted.bind(this));
    this.events.on('report:test:finished', this.testFinished.bind(this));
    this.events.on('report:runner:finished', this.runnerFinished.bind(this));
    this.events.on('report:log:user',this.messageLog.bind(this));
    this.events.on('report:screenshot',this.screenshot.bind(this));
    return this;
  },

  /**
   * Stores the current browser name
   *
   * @method runBrowser
   * @param {string} browser Browser name
   * @chainable
   */

  runBrowser: function (browser) {
    this.browser = browser;
    return this;
  },

  /**
   * Generates JSON for an action
   *
   * @method action
   * @param {object} data Event data
   * @chainable
   */

  action: function (data) {
    data.kind = 'action';
    this.actionQueue.push(data);
    return this;
  },

  /**
   * Generates JSON for an assertion
   *
   * @method assertion
   * @param {object} data Event data
   * @chainable
   */

  assertion: function (data) {
    data.kind = 'assertion';
    this.actionQueue.push(data);
    return this;
  },

  /**
   * Sets up a new testcase
   *
   * @method testStarted
   * @param {object} data Event data
   * @chainable
   */

  testStarted: function (data) {
    this.currentTest = data;
    this.actionQueue = [];
    return this;
  },

  /**
   * Writes data for a finished testcase
   *
   * @method testFinished
   * @param {object} data Event data
   * @chainable
   */

  testFinished: function (data) {
    this.data.tests.push({
      id: data.id,
      name: data.name,
      browser: this.browser,
      status: data.status,
      passedAssertions: data.passedAssertions,
      failedAssertions: data.failedAssertions,
      actions: this.actionQueue
    });
    return this;
  },

  /**
   * Serializes JSON and writes file to the file system
   *
   * @method runnerFinished
   * @param {object} data Event data
   * @chainable
   */

  runnerFinished: function (data) {
    this.data.elapsedTime = data.elapsedTime;
    this.data.status = data.status;
    this.data.assertions = data.assertions;
    this.data.assertionsFailed = data.assertionsFailed;
    this.data.assertionsPassed = data.assertionsPassed;

    var contents = JSON.stringify(this.data, false, 4);

    if (path.extname(this.dest) !== '.json') {
      this.dest = this.dest + '/dalek.json';
    }

    this.events.emit('report:written', {type: 'json', dest: this.dest});
    this._recursiveMakeDirSync(path.dirname(this.dest.replace(path.basename(this.dest, ''))));
    fs.writeFileSync(this.dest, contents, 'utf8');
    return this;
  },

   /**
    * Generates JSON for a message.log
    *
    * @method assertion
    * @param {object} data Event data
    * @chainable
    */

  messageLog: function(data) {
    var logData = {
      kind: 'message',
      message: data
    };
    this.actionQueue.push(logData);
    return this;
  },

  /**
   * Generates JSON for a screenshot
   *
   * @method assertion
   * @param {object} data Event data
   * @chainable
   */

  screenshot : function(data) {
    data.kind = 'screenshot';
    this.actionQueue.push(data);
    return this;
  },


  /**
   * Helper method to generate deeper nested directory structures
   *
   * @method _recursiveMakeDirSync
   * @param {string} path PAth to create
   */

  _recursiveMakeDirSync: function (path) {
    var pathSep = require('path').sep;
    var dirs = path.split(pathSep);
    var root = '';

    while (dirs.length > 0) {
      var dir = dirs.shift();
      if (dir === '') {
        root = pathSep;
      }
      if (!fs.existsSync(root + dir)) {
        fs.mkdirSync(root + dir);
      }
      root += dir + pathSep;
    }
  }
};
